File: R/algorithmic_toolbox/23_knapsack_without_repetition.R
#!/usr/bin/env Rscript
ReadInputData <- function(conn) {
    params <- as.integer(unlist(strsplit(readLines(conn, n = 1), split = " ")))
    names(params) <- c("capacity", "nrWeights")
    weights <- as.integer(unlist(strsplit(readLines(conn, n = 1), split = " ")))
    list(weights=weights, nrWeights=params["nrWeights"], capacity=params["capacity"])
}

SolveKnapsackProblem <- function(weights, nrWeights, capacity) {
    knapsack <- matrix(data = -1, nrow = capacity, ncol = nrWeights)
    solve <- function(currentCapacity, n) {
        if (currentCapacity == 0 || n == 0) return(0)
        if (knapsack[currentCapacity, n] != -1) return(knapsack[currentCapacity, n])
        weight <- weights[n]
        optimalWeight <- if (currentCapacity < weight) solve(currentCapacity, n - 1) else {
            max(solve(currentCapacity - weight, n - 1) + weight, solve(currentCapacity, n - 1))
        }
        knapsack[currentCapacity, n] <- optimalWeight
        optimalWeight
    }
    solve(capacity, nrWeights)
}

Main <- function() {
    conn <- file("stdin", "r")
    inputData <- ReadInputData(conn)
    close(conn)
    result <- SolveKnapsackProblem(inputData$weights, inputData$nrWeights, inputData$capacity)
    cat(result, fill = TRUE)
}

if (!interactive()) {
    Main()
}

File: Haskell/algorithmic_toolbox/23-KnapsackWithoutRepetition.hs
import Control.Monad (replicateM)
import Data.Array (listArray, (!), range)

convertToIntList :: String -> [Int]
convertToIntList = map read . words


readInput :: Int -> IO [[Int]]
readInput n = do
    lines' <- replicateM n getLine
    return $ map convertToIntList lines'


solveKnapsackProblem :: [Int] -> Int -> Int -> Int
solveKnapsackProblem ws n capacity = knapsack capacity n
    where
        weights = listArray (1, n) ws
        knapsack :: Int -> Int -> Int
        knapsack c ix
            | c == 0 || ix == 0 = 0
            | c < weight = solution ! (c, ix - 1)
            | otherwise = max (solution ! (c - weight, ix - 1) + weight) (solution ! (c, ix - 1))
                where
                    weight = weights ! ix
                    bounds = ((0, 0), (capacity, n))
                    solution = listArray bounds [knapsack c k | (c, k) <- range bounds]


main :: IO()
main = do
    [[capacity, nrWeights], weights] <- readInput 2
    print $ solveKnapsackProblem weights nrWeights capacity
